crypto/rsa/rsa_lib.c

```diff
@@ -163,6 +163,139 @@ void *RSA_get_ex_data(const RSA *r, int idx)
+
+/*
+ * Multiply two scale integers together and rescale the result.
+ */
+static ossl_inline uint64_t mul2(uint64_t a, uint64_t b)
+{
+    /*
+     * If correct rounded is desired, enable this define.
+     * Correct rounding being:
+     *      1) if the fractional part is less than one half, round down.
+     *      2) if the fractional part is greater than one half, round up.
+     *      3) if the fractional part is exactly one half, round to even.
+     * This very slightly increases accuracy.
+     */
+#ifdef OSSL_CORRECT_ROUNDING
+    uint64_t m = a * b;
+    uint32_t f = m & (scale - 1);
+
+    m /= scale;
+    if (f == scale / 2)         /* Round to even */
+        m &= ~1;
+    else if (f >= scale / 2)    /* Round up */
+        m++;
+    return m;
+#else
+    return a * b / scale;
+#endif
+}
+
+/*
+ * Calculate the cube root of a 64 bit scaled integer.
+ * Although the cube root of a 64 bit number does fit into a 32 bit unsigned
+ * integer, this is not guaranteed after scaling, so this function has a
+ * 64 bit return.  This uses the shifting nth root algorithm.
+ */
+static uint64_t icbrt64(uint64_t x)
+{
+    uint32_t r = 0;
+    uint64_t b;
+    int s;
+
+    for (s = 63; s >= 0; s -= 3) {
+        r += r;
+        b = 3 * r * (uint64_t)(r + 1) + 1;
+        if ((x >> s) >= b) {
+            x -= b << s;
+            r++;
+        }
+    }
+    return (uint64_t)r * cbrt_scale;
+}
+
+/*
+ * Calculate the natural logarithm of a 64 bit scaled integer.
+ * This is done by calculating a base two logarithm and scaling.
+ * The maximum logarithm (base 2) is 64 and this reduces base e, so
+ * a 32 bit result should not overflow.  The argument passed must be
+ * greater than unity so we don't need to handle negative results.
+ */
+static uint32_t ilog_e(uint64_t v)
+{
+    uint32_t i, r = 0;
+
+    /*
+     * Scale down the value into the range 1 .. 2.
+     *
+     * If fractional numbers need to be processed, another loop needs
+     * to go here that checks v < scale and if so multiplies it by 2 and
+     * reduces r by scale.  This also means making r signed.
+     */
+    while (v >= 2 * scale) {
+        v >>= 1;
+        r += scale;
+    }
+    for (i = scale / 2; i != 0; i /= 2) {
+        v = mul2(v, v);
+        if (v >= 2 * scale) {
+            v >>= 1;
+            r += i;
+        }
+    }
+    r = (r * (uint64_t)scale) / log_e;
+    return r;
+}
+
+/*
+ * NIST SP 800-56B rev 2 Appendix D: Maximum Security Strength Estimates for IFC
+ * Modulus Lengths.
+ *
+ * E = \frac{1.923 \sqrt[3]{nBits \cdot log_e(2)}
+ *           \cdot(log_e(nBits \cdot log_e(2))^{2/3} - 4.69}{log_e(2)}
+ * The two cube roots are merged together here.
+ */
+static uint16_t rsa_compute_security_bits(int n)
+{
+    uint64_t x;
+    uint32_t lx;
+    uint16_t y;
+
+    /*
+     * The first incorrect result (i.e. not accurate or off by one low) occurs
+     * for n = 699668.  The true value here is 1200.  Instead of using this n
```

Comment:
Inaccurate due to rounding? If so why would we ever define OSSL_CORRECT_ROUNDING, since we never get to a value where it would make a difference anyway.